packages/fetchai/connections/ledger_api/connection.py

```diff
@@ -0,0 +1,239 @@
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+
+"""Scaffold connection and channel."""
+import asyncio
+from asyncio import Task
+from collections import deque
+from concurrent.futures import Executor
+from typing import Callable, Deque, List, Optional, cast
+
+import aea
+from aea.configurations.base import ConnectionConfig, PublicId
+from aea.connections.base import Connection
+from aea.crypto.base import LedgerApi
+from aea.crypto.wallet import CryptoStore
+from aea.identity.base import Identity
+from aea.mail.base import Envelope
+
+from packages.fetchai.protocols.ledger_api import LedgerApiMessage
+
+
+class LedgerApiConnection(Connection):
+    """Proxy to the functionality of the SDK or API."""
+
+    connection_id = PublicId.from_str("fetchai/ledger_api:0.1.0")
+
+    def __init__(
+        self,
+        configuration: ConnectionConfig,
+        identity: Identity,
+        crypto_store: CryptoStore,
+    ):
+        """
+        Initialize a connection to interact with a ledger APIs.
+
+        :param configuration: the connection configuration.
+        :param crypto_store: object to access the connection crypto objects.
+        :param identity: the identity object.
+        """
+        super().__init__(
+            configuration=configuration, crypto_store=crypto_store, identity=identity
+        )
+
+        self._dispatcher = _RequestDispatcher(self.loop)
+
+        self.receiving_tasks: List[asyncio.Future] = []
+        self.done_tasks: Deque[asyncio.Future] = deque()
+
+    async def connect(self) -> None:
+        """Set up the connection."""
+
+    async def disconnect(self) -> None:
+        """Tear down the connection."""
+        for task in self.receiving_tasks:
+            if not task.cancelled():
+                task.cancel()
+
+    async def send(self, envelope: "Envelope") -> None:
+        """
+        Send an envelope.
+
+        :param envelope: the envelope to send.
+        :return: None
+        """
+        if isinstance(envelope.message, bytes):
+            message = cast(
+                LedgerApiMessage,
+                LedgerApiMessage.serializer.decode(envelope.message_bytes),
+            )
+        else:
+            message = cast(LedgerApiMessage, envelope.message)
+        api = aea.crypto.registries.make_ledger_api(message.ledger_id)
+        task = self._dispatcher.dispatch(api, message)
+        self.receiving_tasks.append(task)
+
+    async def receive(self, *args, **kwargs) -> Optional["Envelope"]:
+        """
+        Receive an envelope. Blocking.
+
+        :return: the envelope received, or None.
+        """
+        # if there are done tasks, return the result
+        if len(self.done_tasks) > 0:
+            message = self.done_tasks.pop().result()
+            envelope = Envelope(
+                to=self.address,
+                sender="",
```

Comment:
Better to give the name of the ledger api?

Reply:
ok